File: frontend/src/lib/paypal-client.ts
const PAYPAL_API = process.env.NODE_ENV === 'production'
  ? 'https://api-m.paypal.com'
  : 'https://api-m.sandbox.paypal.com';

interface PayPalTokenResponse {
  access_token: string;
  expires_in: number;
}

class PayPalClient {
  private static instance: PayPalClient;
  private accessToken: string | null = null;
  private tokenExpiry: number = 0;
  private apiUrl: string;

  private constructor() {
    this.apiUrl = PAYPAL_API;
  }

  public static getInstance(): PayPalClient {
    if (!PayPalClient.instance) {
      PayPalClient.instance = new PayPalClient();
    }
    return PayPalClient.instance;
  }

  private async refreshAccessToken(): Promise<string> {
    const clientId = process.env.NEXT_PUBLIC_PAYPAL_CLIENT_ID;
    const clientSecret = process.env.PAYPAL_CLIENT_SECRET;

    if (!clientId || !clientSecret) {
      throw new Error('PayPal credentials not configured');
    }

    try {
      const auth = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
      const response = await fetch(`${this.apiUrl}/v1/oauth2/token`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Authorization': `Basic ${auth}`,
        },
        body: 'grant_type=client_credentials',
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`PayPal API error: ${error.error_description || response.statusText}`);
      }

      const data: PayPalTokenResponse = await response.json();
      
      this.accessToken = data.access_token;
      // Set expiry 5 minutes before actual expiry to ensure token validity
      this.tokenExpiry = Date.now() + ((data.expires_in - 300) * 1000);
      
      return this.accessToken;
    } catch (error) {
      console.error('Error refreshing PayPal access token:', error);
      throw new Error('Failed to refresh PayPal access token');
    }
  }

  public async getAccessToken(): Promise<string> {
    if (!this.accessToken || Date.now() >= this.tokenExpiry) {
      return this.refreshAccessToken();
    }
    return this.accessToken;
  }

  public async createOrder(planId: string): Promise<any> {
    try {
      const accessToken = await this.getAccessToken();
      const response = await fetch(`${this.apiUrl}/v1/billing/subscriptions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${accessToken}`,
        },
        body: JSON.stringify({
          plan_id: planId,
          application_context: {
            return_url: `${window.location.origin}/subscription/success`,
            cancel_url: `${window.location.origin}/subscription/cancel`,
          },
        }),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`PayPal order creation failed: ${error.message || response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error creating PayPal order:', error);
      throw error;
    }
  }

  public async cancelSubscription(subscriptionId: string, reason: string): Promise<void> {
    try {
      const accessToken = await this.getAccessToken();
      
      const response = await fetch(`${this.apiUrl}/v1/billing/subscriptions/${subscriptionId}/cancel`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${accessToken}`,
        },
        body: JSON.stringify({ reason }),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`PayPal subscription cancellation failed: ${error.message || response.statusText}`);
      }
    } catch (error) {
      console.error('Error cancelling PayPal subscription:', error);
      throw error;
    }
  }

  public async verifyWebhookSignature(webhookData: any, headers: Headers): Promise<boolean> {
    try {
      const accessToken = await this.getAccessToken();
      const webhookId = process.env.PAYPAL_WEBHOOK_ID;

      if (!webhookId) {
        throw new Error('PayPal webhook ID not configured');
      }

      const response = await fetch(`${this.apiUrl}/v1/notifications/verify-webhook-signature`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${accessToken}`,
        },
        body: JSON.stringify({
          auth_algo: headers.get('paypal-auth-algo'),
          cert_url: headers.get('paypal-cert-url'),
          transmission_id: headers.get('paypal-transmission-id'),
          transmission_sig: headers.get('paypal-transmission-sig'),
          transmission_time: headers.get('paypal-transmission-time'),
          webhook_id: webhookId,
          webhook_event: webhookData,
        }),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`PayPal webhook verification failed: ${error.message || response.statusText}`);
      }

      const data = await response.json();
      return data.verification_status === 'SUCCESS';
    } catch (error) {
      console.error('Error verifying PayPal webhook signature:', error);
      throw error;
    }
  }
}

export const paypalClient = PayPalClient.getInstance(); 

File: backend/app/api/v1/payments/paypal_client.py
import os
import base64
import requests
from typing import Optional

class PayPalClient:
    def __init__(self):
        self.client_id = os.getenv("PAYPAL_CLIENT_ID")
        self.client_secret = os.getenv("PAYPAL_CLIENT_SECRET")
        self.mode = os.getenv("PAYPAL_MODE", "sandbox")
        self.api_url = "https://api-m.sandbox.paypal.com" if self.mode == "sandbox" else "https://api-m.paypal.com"

    async def get_access_token(self) -> str:
        auth = base64.b64encode(f"{self.client_id}:{self.client_secret}".encode()).decode()
        response = requests.post(
            f"{self.api_url}/v1/oauth2/token",
            headers={
                "Authorization": f"Basic {auth}",
                "Content-Type": "application/x-www-form-urlencoded"
            },
            data="grant_type=client_credentials"
        )
        data = response.json()
        return data["access_token"]

    async def create_subscription(self, plan_id: str, user_email: str, user_name: str) -> dict:
        access_token = await self.get_access_token()
        response = requests.post(
            f"{self.api_url}/v1/billing/subscriptions",
            headers={
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json",
                "PayPal-Request-Id": f"{user_email}-{plan_id}-{os.urandom(8).hex()}"
            },
            json={
                "plan_id": plan_id,
                "subscriber": {
                    "name": {
                        "given_name": user_name,
                        "surname": "SoulDream"
                    },
                    "email_address": user_email
                },
                "application_context": {
                    "brand_name": "SoulDream",
                    "locale": "es-ES",
                    "shipping_preference": "NO_SHIPPING",
                    "user_action": "SUBSCRIBE_NOW",
                    "payment_method": {
                        "payer_selected": "PAYPAL",
                        "payee_preferred": "IMMEDIATE_PAYMENT_REQUIRED"
                    },
                    "return_url": f"{os.getenv('FRONTEND_URL')}/subscription/success",
                    "cancel_url": f"{os.getenv('FRONTEND_URL')}/subscription/cancel"
                }
            }
        )
        return response.json()

    async def cancel_subscription(self, subscription_id: str, reason: str) -> None:
        access_token = await self.get_access_token()
        response = requests.post(
            f"{self.api_url}/v1/billing/subscriptions/{subscription_id}/cancel",
            headers={
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json"
            },
            json={"reason": reason}
        )
        if not response.ok:
            response.raise_for_status()

    async def get_subscription_details(self, subscription_id: str) -> dict:
        access_token = await self.get_access_token()
        response = requests.get(
            f"{self.api_url}/v1/billing/subscriptions/{subscription_id}",
            headers={
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json"
            }
        )
        return response.json()

paypal_client = PayPalClient() 